src/core/graph.ts

```diff
@@ -392,6 +392,11 @@ export class SortablePrecedenceGraph<
         this.addNode(leaf);
         search(leaf, pg.getUpstreamNeighbors(leaf));
         visited.clear();
+      } else {
+        // leaf is not a type. It's upstream neighbors become a new leafs.
+        for (const newLeaf of pg.getUpstreamNeighbors(leaf)) {
+          leafs.add(newLeaf);
```

Comment:
This adds entries to a set already being iterated over, and I don't think the iterator gets updated, so I suspect there are corner cases in which this doesn't work. Shouldn't the call to `search` simply be outside of the `if` block? Also, should we really be clearing the `visited` set at all?

Reply:
The intended behavior of this part is like below.
```mermaid
graph
0[port0]
1[reaction0]
2[reaction1]
3[port1]
4[other upstreams...]
5[upstreams...]

1 --> 0
3 --> 0
2 --> 3
4 -->|"search(reaction0, other upstreams...)"| 2
5 -->|"search(reaction0, upstreams...)"| 1
```
Before this modification, `reaction0` and `reaction1` are just ignored as `port0` is `leaf` and is not an instance of `reaction`.

And I agree with
>This adds entries to a set already being iterated over, and I don't think the iterator gets updated, so I suspect there are corner cases in which this doesn't work.

Particularly, in case multiple leaves share the same upstream node. For instance, if we have a dependency graph like below, `port0` is added twice.
```mermaid
graph
0[port0]
1[reaction2]
2[reaction3]
3[port1]
4[other upstreams...]
5[upstreams...]
6[reaction0]
7[reaction1]

1 --> 0
3 --> 0
2 --> 3
4 --> 2
5 --> 1
0 --> 6
0 --> 7
```
Nevertheless, I don't think 
>Shouldn't the call to search simply be outside of the if block? 

will work. Because the first parameter of `search` has to have a type of `T` (`T` is the same as `type`). 
So, how about adding upstreams of the leaf to `leafs` if those are not visited yet?

Also, about the below comment,
>Also, should we really be clearing the visited set at all?

I'm with you that that part is not necessary. However, I'm not 100% sure as I didn't write that code. @axmmisaka Could you please give your opinion about deleting this line? https://github.com/lf-lang/reactor-ts/blob/05480e333401b16149b1b7e7da73bf8bdf6d86ba/src/core/graph.ts#L394

Reply:
I think I may have written that code :rofl:

I pushed some changes to clarify what the code does...